### DISCUSSION TITLE
Shelly EM activate 365Days Energie History (Data Saving) in Tasmota ?

### DISCUSSION THREAD

**Discussion by Starfoxfs** [arendst/Tasmota#17001] [Q&A] [answered]
Hi,

i have bought a new Shelly EM and flashed it with newest Tasmota Release (OTA).

How can i activate the 365 days Energie History saving in Tasmota to Store Data like Today, Yesterday and Total if i have a Powerloss ?

At the moment when i disable Power for the Shelly Device the Total, Today and Yesterday Data would be reseted to 0.

My actual Template is:

`{"NAME":"Shelly EM","GPIO":[0,0,0,0,0,0,0,0,640,3457,608,224,8832,1],"FLAG":0,"BASE":18}`

and my Tasmota Status 0 shows this:

`18:27:00.894 CMD: Status 0
18:27:00.901 MQT: stat/tasmota_7456F9/STATUS = {"Status":{"Module":0,"DeviceName":"ShellyEM","FriendlyName":["ShellyEM"],"Topic":"tasmota_7456F9","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
18:27:00.908 MQT: stat/tasmota_7456F9/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Power On","Uptime":"0T00:13:30","StartupUTC":"2022-11-05T17:13:30","Sleep":50,"CfgHolder":4617,"BootCount":27,"BCResetTime":"2022-11-04T17:54:41","SaveCount":237,"SaveAddress":"F7000"}}
18:27:00.915 MQT: stat/tasmota_7456F9/STATUS2 = {"StatusFWR":{"Version":"12.2.0(tasmota)","BuildDateTime":"2022-10-17T08:39:23","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"372/699"}}
18:27:00.921 MQT: stat/tasmota_7456F9/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["xxx",""],"TelePeriod":60,"Resolution":"558180C0","SetOption":["00208009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
18:27:00.935 MQT: stat/tasmota_7456F9/STATUS4 = {"StatusMEM":{"ProgramSize":626,"Free":376,"Heap":24,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"1620C2","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
18:27:00.943 MQT: stat/tasmota_7456F9/STATUS5 = {"StatusNET":{"Hostname":"tasmota-7456F9-5881","IPAddress":"x.x.x.x","Gateway":"x.x.x.x","Subnetmask":"x.x.x.x","DNSServer1":"x.x.x.x","DNSServer2":"0.0.0.0","Mac":"34:94:54:74:56:F9","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
18:27:00.949 MQT: stat/tasmota_7456F9/STATUS6 = {"StatusMQT":{"MqttHost":"xxx","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_7456F9","MqttUser":"xxx","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:27:00.958 MQT: stat/tasmota_7456F9/STATUS7 = {"StatusTIM":{"UTC":"2022-11-05T17:27:00","Local":"2022-11-05T18:27:00","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"07:43","Sunset":"17:23"}}
18:27:00.964 MQT: stat/tasmota_7456F9/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
18:27:00.976 MQT: stat/tasmota_7456F9/STATUS10 = {"StatusSNS":{"Time":"2022-11-05T18:27:00","ENERGY":{"TotalStartTime":"2022-11-04T17:54:41","Total":0.001,"Yesterday":0.000,"Today":0.001,"TodaySumImport":0.000,"TodaySumExport":0.000,"ExportActive":[0.000,0.000],"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":50,"Voltage":232,"Current":[0.000,0.000]}}}
18:27:00.985 MQT: stat/tasmota_7456F9/STATUS11 = {"StatusSTS":{"Time":"2022-11-05T18:27:00","Uptime":"0T00:13:30","UptimeSec":810,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"xxx","BSSId":"44:4E:6D:A3:45:E7","Channel":6,"Mode":"11n","RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:03"}}}`

**Comment by barbudor** [comment, marked as answer]
Strictly, there no such feature
However you can use a rule to force saving periodically at the cost of wearing the flash faster
On tele-Energy do savedata endon
Would do that at each teleperiod

**Reply by sfromis** [reply]
What I do is to have my backend record the sensor payloads arriving via MQTT, providing me lossless (except for part of one TelePeriod) data, available for statistics with details, without extra flash wear.

**Comment by Starfoxfs** [comment]
Ok thanks to all, i have MQTT Backend with NodeRed so i make some Nodes to save it to Influx Database

**Comment by Starfoxfs** [comment]
Update:
It works now from itself ?!?
I don´t have switched anything, last Night the Shelly EM was Powerless and now i see the 0,12 kwh Total and Yesterday 0,12 kwh.

In the description of the Shelly EM it says there is an internal Storage for 365days in case Wifi turns off.

I think Tasmota uses this internal Storage automaticly ?!?

**Reply by sfromis** [reply]
That "internal storage" does not exist as as hardware feature, such is entirely up to the firmware. And Tasmota works in very different ways from the factory firmware, and does not try to mimic the firmware features of some "random" commercial device. In Tasmota, energy consumption data is saved along with settings data, meaning that values will be saved when something happens to cause saving, by default including when switching output relays on/off, or doing setup changes. And data will be saved at midnight, if running at that time.

It was a design decision for Tasmota to not save data all the time, to the possible detriment of device lifetime (maybe the Shelly firmware has much wider wear leveling to mitigate this). With energy monitoring, cutting off input power to the device, without any chance of a save being triggered, is not a good strategy with Tasmota. My energy monitoring devices are on 24/7.

**Comment by Starfoxfs** [comment]
Thanks for the detailed explanation, it is also not my intention to switch off the device manually, only in the event of a power failure should the total values ​​(like an electricity meter) be retained.

**Reply by sfromis** [reply]
With a sudden power cut, it will not retain energy consumption added since last save of the settings data block, whatever way that save is triggered. Your backend can do that, up to the last TelePeriod message with the Total and Today values up to date. You could create an automation to push these values back to Tasmota after restart, if you want long-time values in Tasmota to be "as correct as possible"
